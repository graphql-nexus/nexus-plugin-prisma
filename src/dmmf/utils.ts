import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
