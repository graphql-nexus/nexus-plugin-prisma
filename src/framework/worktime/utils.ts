import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
