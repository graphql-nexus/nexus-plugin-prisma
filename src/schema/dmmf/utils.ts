import { DMMF } from '@prisma/client/runtime'

export const getPrismaClientDmmf = (packagePath: string): DMMF.Document => {
  let dmmf: undefined | DMMF.Document = undefined

  try {
    dmmf = require(packagePath).dmmf
  } catch (error) {
    throw new Error(
      `Failed to import prisma client package at ${packagePath}. The following error occured while trying:
        ${error.stack}`
    )
  }

  if (!dmmf) {
    throw new Error(`\
You most likely forgot to initialize the Prisma Client. Please run \`prisma generate\` and try to run it again.
If that does not solve your problem, please open an issue.`)
  }

  return dmmf
}
