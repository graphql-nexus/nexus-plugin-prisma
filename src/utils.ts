import { relative } from 'path'
import { ExternalDMMF as DMMF } from './dmmf/dmmf-types'
import {
  OperationName,
  IFieldNamingStrategy,
} from './nexus-prisma/NamingStrategies'
import { core } from 'nexus'

export const keyBy: <T>(
  collection: T[],
  iteratee: (value: T) => string,
) => Record<string, T> = (collection, iteratee) => {
  return collection.reduce<any>((acc, curr) => {
    acc[iteratee(curr)] = curr
    return acc
  }, {})
}

export function partition<T>(
  arr: T[],
  iteratee: (val: T) => boolean,
): [T[], T[]] {
  const partitioned: [T[], T[]] = [[], []]

  for (const val of arr) {
    const partitionIndex: 0 | 1 = iteratee(val) ? 0 : 1
    partitioned[partitionIndex].push(val)
  }

  return partitioned
}

export const upperFirst = (s: string): string => {
  return s.replace(/^\w/, c => c.toUpperCase())
}

export function flatMap<T, U>(
  array: T[],
  callbackfn: (value: T, index: number, array: T[]) => U[],
): U[] {
  return Array.prototype.concat(...array.map(callbackfn))
}

export function nexusOpts(param: {
  type: core.AllInputTypes | core.AllNexusInputTypeDefs<string>
  isList: boolean
  isRequired: boolean
}): {
  type: any
  list: true | undefined
  nullable: boolean
} {
  return {
    type: param.type,
    list: param.isList ? true : undefined,
    nullable: !param.isRequired,
  }
}

export function assertPhotonInContext(photon: any) {
  if (!photon) {
    throw new Error('Could not find photon in context')
  }
}

export function trimIfInNodeModules(path: string) {
  if (path.includes('node_modules')) {
    return path.substring(
      path.lastIndexOf('node_modules') + 'node_modules'.length + 1,
    )
  }

  return path
}

export function getImportPathRelativeToOutput(
  from: string,
  to: string,
): string {
  if (to.includes('node_modules')) {
    return trimIfInNodeModules(to)
  }

  let relativePath = relative(from, to)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}

export function getCRUDFieldName(
  modelName: string,
  fieldName: string,
  mapping: DMMF.Mapping,
  namingStrategy: IFieldNamingStrategy,
) {
  const operationName = Object.keys(mapping).find(
    key => (mapping as any)[key] === fieldName,
  ) as OperationName | undefined

  if (!operationName || !namingStrategy[operationName]) {
    throw new Error(`Could not find mapping for field ${fieldName}`)
  }

  return namingStrategy[operationName](fieldName, modelName)
}
