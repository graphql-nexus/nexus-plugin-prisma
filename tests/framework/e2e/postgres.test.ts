import { createE2EContext } from 'nexus/dist/lib/e2e-testing'
import { getTmpDir } from 'nexus/dist/lib/fs'
import * as Path from 'path'
import { refCount } from 'rxjs/operators'
import { e2eTestPlugin } from '../__helpers/e2e/testing'
import { bufferOutput, takeUntilServerListening } from '../__helpers/e2e/utils'

const tmpDir = getTmpDir()
const ctx = createE2EContext({
  dir: Path.join(tmpDir, 'postgres'),
  localNexus: null,
  serverPort: 4000,
})

test('e2e', async () => {
  let nexusVersion = process.env.NEXUS_VERSION ?? 'next'

  if (!process.env.NEXUS_PLUGIN_PRISMA_VERSION) {
    throw new Error('env var NEXUS_PLUGIN_PRISMA_VERSION must be set')
  }

  // Run npx nexus from local path
  const initResult = await ctx
    .npxNexusCreateApp({
      packageManagerType: 'npm',
      databaseType: 'PostgreSQL',
      prismaPluginVersion: process.env.NEXUS_PLUGIN_PRISMA_VERSION,
      nexusVersion,
    })
    .pipe(refCount(), takeUntilServerListening, bufferOutput)
    .toPromise()

  expect(initResult).toContain('Run `yarn -s dev` to start working')

  await e2eTestPlugin(ctx)
})
