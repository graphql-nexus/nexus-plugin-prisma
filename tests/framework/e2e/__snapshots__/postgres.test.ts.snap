// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e: introspection 1`] = `
Object {
  "__schema": Object {
    "directives": Array [
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "Included when true.",
            "name": "if",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
        ],
        "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "include",
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "Skipped when true.",
            "name": "if",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
        ],
        "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "skip",
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": "\\"No longer supported\\"",
            "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax (as specified by [CommonMark](https://commonmark.org/).",
            "name": "reason",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "description": "Marks an element of a GraphQL schema as no longer supported.",
        "locations": Array [
          "FIELD_DEFINITION",
          "ENUM_VALUE",
        ],
        "name": "deprecated",
      },
    ],
    "mutationType": null,
    "queryType": Object {
      "name": "Query",
    },
    "subscriptionType": null,
    "types": Array [
      Object {
        "description": null,
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "defaultValue": null,
                "description": null,
                "name": "world",
                "type": Object {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "hello",
            "type": Object {
              "kind": "OBJECT",
              "name": "World",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "worlds",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "World",
                  "ofType": null,
                },
              },
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "Query",
        "possibleTypes": null,
      },
      Object {
        "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "String",
        "possibleTypes": null,
      },
      Object {
        "description": null,
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Int",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "population",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Float",
                "ofType": null,
              },
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "World",
        "possibleTypes": null,
      },
      Object {
        "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1.",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "Int",
        "possibleTypes": null,
      },
      Object {
        "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](https://en.wikipedia.org/wiki/IEEE_floating_point).",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "Float",
        "possibleTypes": null,
      },
      Object {
        "description": "A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all available types and directives on the server, as well as the entry points for query, mutation, and subscription operations.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A list of all types supported by this server.",
            "isDeprecated": false,
            "name": "types",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "NON_NULL",
                  "name": null,
                  "ofType": Object {
                    "kind": "OBJECT",
                    "name": "__Type",
                    "ofType": null,
                  },
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The type that query operations will be rooted at.",
            "isDeprecated": false,
            "name": "queryType",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "OBJECT",
                "name": "__Type",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "If this server supports mutation, the type that mutation operations will be rooted at.",
            "isDeprecated": false,
            "name": "mutationType",
            "type": Object {
              "kind": "OBJECT",
              "name": "__Type",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "If this server support subscription, the type that subscription operations will be rooted at.",
            "isDeprecated": false,
            "name": "subscriptionType",
            "type": Object {
              "kind": "OBJECT",
              "name": "__Type",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A list of all directives supported by this server.",
            "isDeprecated": false,
            "name": "directives",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "NON_NULL",
                  "name": null,
                  "ofType": Object {
                    "kind": "OBJECT",
                    "name": "__Directive",
                    "ofType": null,
                  },
                },
              },
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__Schema",
        "possibleTypes": null,
      },
      Object {
        "description": "The fundamental unit of any GraphQL Schema is the type. There are many kinds of types in GraphQL as represented by the \`__TypeKind\` enum.

Depending on the kind of a type, certain fields describe information about that type. Scalar types provide no information beyond a name and description, while Enum types provide their values. Object and Interface types provide the fields they describe. Abstract types, Union and Interface, provide the Object types possible at runtime. List and NonNull types compose other types.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "kind",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "ENUM",
                "name": "__TypeKind",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "description",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [
              Object {
                "defaultValue": "false",
                "description": null,
                "name": "includeDeprecated",
                "type": Object {
                  "kind": "SCALAR",
                  "name": "Boolean",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "fields",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "__Field",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "interfaces",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "__Type",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "possibleTypes",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "__Type",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [
              Object {
                "defaultValue": "false",
                "description": null,
                "name": "includeDeprecated",
                "type": Object {
                  "kind": "SCALAR",
                  "name": "Boolean",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "enumValues",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "__EnumValue",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "inputFields",
            "type": Object {
              "kind": "LIST",
              "name": null,
              "ofType": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "__InputValue",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "ofType",
            "type": Object {
              "kind": "OBJECT",
              "name": "__Type",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__Type",
        "possibleTypes": null,
      },
      Object {
        "description": "An enum describing what kind of type a given \`__Type\` is.",
        "enumValues": Array [
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is a scalar.",
            "isDeprecated": false,
            "name": "SCALAR",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is an object. \`fields\` and \`interfaces\` are valid fields.",
            "isDeprecated": false,
            "name": "OBJECT",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is an interface. \`fields\` and \`possibleTypes\` are valid fields.",
            "isDeprecated": false,
            "name": "INTERFACE",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is a union. \`possibleTypes\` is a valid field.",
            "isDeprecated": false,
            "name": "UNION",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is an enum. \`enumValues\` is a valid field.",
            "isDeprecated": false,
            "name": "ENUM",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is an input object. \`inputFields\` is a valid field.",
            "isDeprecated": false,
            "name": "INPUT_OBJECT",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is a list. \`ofType\` is a valid field.",
            "isDeprecated": false,
            "name": "LIST",
          },
          Object {
            "deprecationReason": null,
            "description": "Indicates this type is a non-null. \`ofType\` is a valid field.",
            "isDeprecated": false,
            "name": "NON_NULL",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "ENUM",
        "name": "__TypeKind",
        "possibleTypes": null,
      },
      Object {
        "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "Boolean",
        "possibleTypes": null,
      },
      Object {
        "description": "Object and Interface types are described by a list of Fields, each of which has a name, potentially a list of arguments, and a return type.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "description",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "args",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "NON_NULL",
                  "name": null,
                  "ofType": Object {
                    "kind": "OBJECT",
                    "name": "__InputValue",
                    "ofType": null,
                  },
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "type",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "OBJECT",
                "name": "__Type",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "isDeprecated",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "deprecationReason",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__Field",
        "possibleTypes": null,
      },
      Object {
        "description": "Arguments provided to Fields or Directives and the input fields of an InputObject are represented as Input Values which describe their type and optionally a default value.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "description",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "type",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "OBJECT",
                "name": "__Type",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A GraphQL-formatted string representing the default value for this input value.",
            "isDeprecated": false,
            "name": "defaultValue",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__InputValue",
        "possibleTypes": null,
      },
      Object {
        "description": "One possible value for a given Enum. Enum values are unique values, not a placeholder for a string or numeric value. However an Enum value is returned in a JSON response as a string.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "description",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "isDeprecated",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "deprecationReason",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__EnumValue",
        "possibleTypes": null,
      },
      Object {
        "description": "A Directive provides a way to describe alternate runtime execution and type validation behavior in a GraphQL document.

In some cases, you need to provide options to alter GraphQL's execution behavior in ways field arguments will not suffice, such as conditionally including or skipping a field. Directives provide this by describing additional information to the executor.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "description",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "locations",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "NON_NULL",
                  "name": null,
                  "ofType": Object {
                    "kind": "ENUM",
                    "name": "__DirectiveLocation",
                    "ofType": null,
                  },
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "args",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "NON_NULL",
                  "name": null,
                  "ofType": Object {
                    "kind": "OBJECT",
                    "name": "__InputValue",
                    "ofType": null,
                  },
                },
              },
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "__Directive",
        "possibleTypes": null,
      },
      Object {
        "description": "A Directive can be adjacent to many parts of the GraphQL language, a __DirectiveLocation describes one such possible adjacencies.",
        "enumValues": Array [
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a query operation.",
            "isDeprecated": false,
            "name": "QUERY",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a mutation operation.",
            "isDeprecated": false,
            "name": "MUTATION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a subscription operation.",
            "isDeprecated": false,
            "name": "SUBSCRIPTION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a field.",
            "isDeprecated": false,
            "name": "FIELD",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a fragment definition.",
            "isDeprecated": false,
            "name": "FRAGMENT_DEFINITION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a fragment spread.",
            "isDeprecated": false,
            "name": "FRAGMENT_SPREAD",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an inline fragment.",
            "isDeprecated": false,
            "name": "INLINE_FRAGMENT",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a variable definition.",
            "isDeprecated": false,
            "name": "VARIABLE_DEFINITION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a schema definition.",
            "isDeprecated": false,
            "name": "SCHEMA",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a scalar definition.",
            "isDeprecated": false,
            "name": "SCALAR",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an object type definition.",
            "isDeprecated": false,
            "name": "OBJECT",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a field definition.",
            "isDeprecated": false,
            "name": "FIELD_DEFINITION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an argument definition.",
            "isDeprecated": false,
            "name": "ARGUMENT_DEFINITION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an interface definition.",
            "isDeprecated": false,
            "name": "INTERFACE",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to a union definition.",
            "isDeprecated": false,
            "name": "UNION",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an enum definition.",
            "isDeprecated": false,
            "name": "ENUM",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an enum value definition.",
            "isDeprecated": false,
            "name": "ENUM_VALUE",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an input object type definition.",
            "isDeprecated": false,
            "name": "INPUT_OBJECT",
          },
          Object {
            "deprecationReason": null,
            "description": "Location adjacent to an input object field definition.",
            "isDeprecated": false,
            "name": "INPUT_FIELD_DEFINITION",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "ENUM",
        "name": "__DirectiveLocation",
        "possibleTypes": null,
      },
      Object {
        "description": "A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "DateTime",
        "possibleTypes": null,
      },
      Object {
        "description": "The \`JSON\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).",
        "enumValues": null,
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "SCALAR",
        "name": "Json",
        "possibleTypes": null,
      },
    ],
  },
}
`;

exports[`e2e: worlds-query 1`] = `
Object {
  "worlds": Array [
    Object {
      "id": 1,
      "name": "Earth",
      "population": 6000000000,
    },
    Object {
      "id": 2,
      "name": "Mars",
      "population": 0,
    },
  ],
}
`;
